### PR TITLE
Re-factored the Instrument class to use an internal definition list for…

### LINKED ISSUE #3: Make the config file be the authority
## Current Behavior

The main internal structure for Guarneri is like:

```python
{
    "Motor": [  # device class
        {
            "name": "m1",
            "prefix": "25idc:m1",
        }
    ]
}
```

Right now, Guarneri iterates through it's mapping of device class names to device classes and checks if that name is present in the internal dictionary.

## New Behavior

Based on discussion with @MDecarabas and @prjemian, we should flip this. Instead, let's create an internal structure from the TOML file that looks something like:

```python
{
    "m1": [
        {
            "device_class": "Motor",
            "kwargs": {
                ...
            }
        }
    ]
}
```
then we can iterate through this structure and construct the devices accordingly.

## Benefits

1) Mistyped classes can be identified when parsing, and raise exceptions.
2) Device class aliases can be optional. If a device class is not in the initial mapping, but is still importable by Guarneri, it can still be added. No coding necessary.

### PR DESCRIPTION
**Pull request by canismarko**
… creating devices.

## Description

This PR implements #3 and #4. It refactors the way config files are parsed to make the Instrument customizable.

## Motivation and Context

Fixes #3 and #4 

## How Has This Been Tested?

Tests have been added. They have been run locally against ophyd-async 0.7.0 and 0.8.0a4. Tests include both threaded Ophyd and ophyd-async devices.

## Where Has This Been Documented?

Documention has been added.

### CODE REVIEW

**Comment by canismarko** [comment]
@prjemian, the file custom_schema.rst describes the way you could use this for your own YAML files. Does that work?

**Comment by prjemian** [comment]
I'll try this out.  I'm sure it works for the standard device creation.

Wonder how it can be used with [`apstools.devices.ad_creator()`](https://bcda-aps.github.io/apstools/latest/api/_devices.html#apstools.devices.area_detector_factory.ad_creator).  That's probably a separate case.

**Comment by canismarko** [comment]
> I'll try this out. I'm sure it works for the standard device creation.
> 
> Wonder how it can be used with [`apstools.devices.ad_creator()`](https://bcda-aps.github.io/apstools/latest/api/_devices.html#apstools.devices.area_detector_factory.ad_creator). That's probably a separate case.

Should work exactly the same, actually. I had a similar need so I designed it so that factory functions work identically to straight Device classes. I haven't written the docs for that yet, but here's the test showing that behavior:

https://github.com/spc-group/guarneri/blob/main/src/guarneri/tests/test_instrument.py#L44

```python
def load_devices(num_devices=0):
    """A simple factory function that loads a bunch of dummy devices."""
    for i in num_devices:
        yield AsyncDevice()

@pytest.fixture()
def instrument():
    inst = Instrument(
        {
            "async_device": AsyncDevice,
            "factory_device": load_devices,
            "threaded_device": ThreadedDevice,
        }
    )
    with open(toml_file, mode="tr", encoding="utf-8") as fd:
        inst.parse_toml_file(fd)
    return inst
```

Caveat: I'm not sure what the parameter *ad_setup* is expected to be since it's type is ``object``. ``Instrument`` shouldn't care, but you'll need to make sure the values fit into the markup language you're using (YAML?).

**Comment by canismarko** [comment]
> My specification for `ad_creator` object is different from the standard spec. For plugin specs that accept the default class provided by the factory, it's not necessary to name that plugin class explicitly (and this might be a challenge to fit the guarneri design). Here's my example:
> 
> ```yaml
> ad_creator:
> 
>   adsimdet:
>       # IOC host: arf
>       allow_warmup: true
>       prefix: "ad:"
>       plugins:
>         cam:
>           class: apstools.devices.SimDetectorCam_V34
>         image: null
>         pva: null
> 
>         hdf1:
>           class: apstools.devices.AD_EpicsFileNameHDF5Plugin
>           kwargs:
>             read_path_template: "/mnt/iocad/tmp/"
>             write_path_template: "/tmp/"
>         roi1: null
>         stats1: null
> ```
> 
> This spec results in the python code:
> 
> ```python
> from apstools.devices import ad_creator
> 
> adsimdet = ad_creator(
>     "ad:",
>     name="adsimdet",
>     plugins=[
>         {"cam": {"class: "apstools.devices.SimDetectorCam_V34"}},
>         "image,
>         "pva",
>         {
>             "hdf1": {
>                 "class": "apstools.devices.AD_EpicsFileNameHDF5Plugin",
>                 "kwargs": {
>                     "read_path_template": "/mnt/iocad/tmp/",
>                     "write_path_template": "/tmp/",
>                 },
>             }
>         },
>         "roi1",
>         "stats1",
>     ]
> )
> ```

Guarneri shouldn't care. It inspects the signature of the Device or factory, makes sure the arguments are of the correct type and that no required arguments are missing, then calls the Device or factory with those arguments. As long as the factory returns a single Device instance or a sequence of Device instances, it *should* work.

Maybe give it a try and let me know how it's not working?

**Comment by prjemian** [comment]
I'll try that.

**Comment by prjemian** [comment]
First problem: `AttributeError` at this line: https://github.com/spc-group/guarneri/blob/25dd1c00952e64fb612a4b459921e226519a2e43/src/guarneri/instrument.py#L61

I'm passing a `str` but the method does not substitute to a `Path` I understand the annotation to say (`| str`).

```py
In [26]: instr = guarneri.Instrument({})

In [27]: instr.parse_config("src/instrument/configs/guarneri-style.yml")
AttributeError: 'str' object has no attribute 'suffix'
```

**Comment by canismarko** [comment]
> First problem: `AttributeError` at this line:

Right you are. Should be fixed now. Thanks.

**Comment by prjemian** [comment]
Some device classes are tripping a `KeyError: Missing required key 'args'` from this method: https://github.com/spc-group/guarneri/blob/25dd1c00952e64fb612a4b459921e226519a2e43/src/guarneri/instrument.py#L174

That method is called here: https://github.com/spc-group/guarneri/blob/25dd1c00952e64fb612a4b459921e226519a2e43/src/guarneri/instrument.py#L154
and here: https://github.com/spc-group/guarneri/blob/25dd1c00952e64fb612a4b459921e226519a2e43/src/guarneri/instrument.py#L159

How can `args` be validated when it is not passed to the validator?

**Comment by prjemian** [comment]
Error is reproducible in TOML.  Config file content:

```toml
[[ SimulatedApsPssShutterWithStatus ]]
name = "shutter"
```

Code that reproduces the exception trace:

```py
"""docs"""

import guarneri
from apstools.devices import SimulatedApsPssShutterWithStatus

xref = {
    "SimulatedApsPssShutterWithStatus": SimulatedApsPssShutterWithStatus,
}
instr = guarneri.Instrument(xref)
instr.load("src/instrument/configs/guarneri-shutters.toml")
print(f"{sorted(instr.devices.device_names)=!r}")
print(f"component count: {len(instr.devices.component_names)}")
```

The exception trace:

```py
(bluesky_2024_3) prjemian@arf:~/bluesky$ python ./dev_demo.py 
Traceback (most recent call last):
  File "/home/prjemian/bluesky/./dev_demo.py", line 10, in <module>
    instr.load("src/instrument/configs/guarneri-shutters.toml")
  File "/home/prjemian/Documents/projects/spc-group/guarneri/src/guarneri/instrument.py", line 351, in load
    devices = self.make_devices(device_defns, fake=fake)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/prjemian/Documents/projects/spc-group/guarneri/src/guarneri/instrument.py", line 154, in make_devices
    self.validate_params(defn["kwargs"], Klass)
  File "/home/prjemian/Documents/projects/spc-group/guarneri/src/guarneri/instrument.py", line 189, in validate_params
    raise InvalidConfiguration(
guarneri.exceptions.InvalidConfiguration: Missing required key 'args' for <class 'apstools.devices.shutters.SimulatedApsPssShutterWithStatus'>: {'name': 'shutter'}
```

**Comment by prjemian** [comment]
This branch is working for me with a custom YAML loader, noting these additional issues:

- https://github.com/BCDA-APS/apstools/issues/1039
- #9
- #6

Here's a fragment sampler of devices from my YAML device file:

```yaml
instrument.devices.ioc_stats.IocInfoDevice: [{name: gp_stats, prefix: "gp:"}]

instrument.devices.kohzu_monochromator.KohzuDCM:
  - name: dcm
    prefix: "gp:"
    labels: ["monochromator"]
    m_th: m45
    m_y: m46
    m_z: m47

apstools.synApps.Optics2Slit2D_HV:
  - name: slit1
    prefix: gp:Slit1
    labels: ["slits"]

apstools.synApps.TransformRecord: [{name: ad_transform, prefix: "gp:userTran19"}]

hkl.SimulatedE4CV:
  - name: sim4c
    prefix: ""
    labels: ["diffractometer"]

ophyd.EpicsMotor:
  - { name: m1, prefix: gp:m1 }
  - { name: m2, prefix: gp:m2 }
  - { name: m3, prefix: gp:m3 }
  - { name: m4, prefix: gp:m4 }

ophyd.EpicsSignalRO:
  - name: noisy
    read_pv: gp:userCalc1
    labels: ["detectors", "simulator"]

ophyd.scaler.ScalerCH:
  - name: scaler1
    prefix: gp:scaler1
    labels: ["scalers", "detectors"]

ophyd.Signal:
  - name: demo_signal
    value: 1.0
    kind: config

apstools.devices.ad_creator:

  - name: adsimk
    # IOC host: arf
    prefix: "kad:"
    labels: ["area_detector", "detectors"]
    plugins:
      - cam:
          class: apstools.devices.SimDetectorCam_V34
      - image
      - pva
      - hdf1:
          class: apstools.devices.AD_EpicsFileNameHDF5Plugin
          read_path_template: "/mnt/iocad/tmp/"
          write_path_template: "/tmp/"
      - roi1
      - stats1
```

Time to create a full configuration of devices (two examples):

```py
sorted(oregistry.device_names)=['ad_transform', 'adsimdet', 'adsimk', 'dcm', 'demo_signal', 'fourc', 'gp_stats', 'm1', 'm10', 'm11', 'm12', 'm13', 'm14', 'm15', 'm16', 'm2', 'm3', 'm4', 'm7', 'm8', 'm9', 'noisy', 'scaler1', 'scan_id_epics', 'sim4c', 'sim6c', 'simk4c', 'simk6c', 'sixc', 'slit1', 'user_calcouts', 'user_calcs', 'user_sseqs', 'user_transforms']
component count: 6470
device count: 34
8.080 s

sorted(oregistry.device_names)=['ad_transform', 'adsimdet', 'adsimk', 'dcm', 'demo_signal', 'fourc', 'gp_stats', 'm1', 'm10', 'm11', 'm12', 'm13', 'm14', 'm15', 'm16', 'm2', 'm3', 'm4', 'm7', 'm8', 'm9', 'noisy', 'scaler1', 'scan_id_epics', 'sim4c', 'sim6c', 'simk4c', 'simk6c', 'sixc', 'slit1', 'user_calcouts', 'user_calcs', 'user_sseqs', 'user_transforms']
component count: 6470
device count: 34
6.586 s
```

**Review by prjemian — APPROVED** [review]

**Comment by prjemian** [comment]
FYI: My custom YAML loader:

```py
def get_object_class(full_path: str) -> type:
    """
    Return the object class given the full import path.

    TODO: hoist to apstools for ad_creator() in https://github.com/BCDA-APS/apstools/issues/1039
    """
    from importlib import import_module

    class_object = None

    module_name, class_name = full_path.rsplit(".", 1)
    module_object = import_module(module_name)
    class_object = getattr(module_object, class_name)

    return class_object


    class Instrument(guarneri.Instrument):
        def parse_config(self, config_file: Path | str) -> list[dict]:
            if config_file.suffix == ".toml":
                return self.parse_toml_file(config_file)
            if config_file.suffix.lower() in (".yaml", ".yml"):
                return self.parse_yaml_file(config_file)
            else:
                raise ValueError(f"Unknown file extension: {config_file}")

        def parse_yaml_file(self, config_file: Path | str) -> list[dict]:
            if isinstance(config_file, str):
                config_file = Path(config_file)

            def parse(class_name, specs):
                if class_name not in self.device_classes:
                    self.device_classes[class_name] = get_object_class(class_name)
                entries = []
                for table in specs:
                    args = table.pop("args", ())
                    entry = {
                        "device_class": class_name,
                        "args": args,
                        "kwargs": table,
                    }
                    entries.append(entry)
                return entries

            controls = load_config_yaml(config_file)
            devices = []
            for k, v in controls.items():
                devices += parse(k, v)
            # print(f"{devices=}")
            return devices
```

File: src/guarneri/instrument.py [hunk 1]
```diff
@@ -50,49 +49,127 @@ class Instrument:
...
-        self.beamline_name = beamline.get("name", self.beamline_name)
-        self.hardware_is_present = beamline.get(
-            "hardware_is_present", self.hardware_is_present
-        )
-        # Make devices from config file
-        return self.parse_config(config)
+        config_file = Path(config_file)
+        if config_file.suffix == ".toml":
+            return self.parse_toml_file(config_file)
+        if config_file.suffix == ".yaml":
```
Comment (prjemian) [inline src/guarneri/instrument.py:99]: It's common for YAML files to use either `.yaml` or `.yml` extensions.  Code should recognize both.

```suggestion
        if config_file.suffix in (".yaml", ".yml"):
```

**Review by prjemian — APPROVED** [review]
With the suggested change below, this works for me.

File: docs/conf.py [hunk 2]
```diff
@@ -46,6 +46,7 @@
     "mergedeep",
 ]
 napoleon_numpy_docstring = True
+pygments_style = "sphinx"
```
Comment (prjemian) [inline docs/conf.py:49]: Why are there two `conf.py` files?  Will this file ever be used by Sphinx?

Comment (canismarko) [reply docs/conf.py:49]: Best I can tell, no that file doesn't ever get used. I copied this cookiecutter from another repository, and I suspect this is cruft.

I've opened an issue for it: #10

**Pull request merged**